File: src/games/stendhal/common/constants/SoundLayer.java
/***************************************************************************
 *                   (C) Copyright 2003-2013 - Stendhal                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.common.constants;

/**
 * the layer on which a sound is played
 *
 * @author hendrik
 */
public enum SoundLayer {

	/** nice background music which is considered out-of-game */
	BACKGROUND_MUSIC("music"),

	/**
	 * sounds that are related to the zone and are in-game,
	 * like a waterfall, wind, the hammering at the blacksmith
	 */
	AMBIENT_SOUND("ambient"),

	/** noise made by creatures and NPCs */
	CREATURE_NOISE("creature"),

	/** noise made by weapons and armor */
	FIGHTING_NOISE("sfx"),

	/** user interface feedback, opening of windows, private message notification */
	USER_INTERFACE("gui");

	/** the internal name of the sound layer */
	public final String groupName;

	/**
	 * creates an instance of SoundLayer
	 *
	 * @param groupName name of group
	 */
	private SoundLayer(String groupName) {
		this.groupName = groupName;
	}
}

File: src/js/stendhal/data/sound/SoundLayer.ts
/***************************************************************************
 *                    Copyright © 2024 - Faiumoni e. V.                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { AbstractEnum } from "../enum/AbstractEnum";


/**
 * Available layers of sound.
 *
 * string names should match `games.stendhal.common.constants.SoundLayer`
 */
export class SoundLayer extends AbstractEnum<string> {

	/** Reference to layers. */
	private static readonly layers: SoundLayer[] = [];

	static readonly MUSIC = new SoundLayer("music");
	static readonly AMBIENT = new SoundLayer("ambient");
	static readonly CREATURE = new SoundLayer("creature");
	static readonly SFX = new SoundLayer("sfx");
	static readonly GUI = new SoundLayer("gui");


	private constructor(value: string) {
		super(value);
		SoundLayer.layers.push(this);
	}

	/**
	 * Retrieves all layer names.
	 *
	 * @return {string[]}
	 *   Names of available layers.
	 */
	static names(): string[] {
		const names: string[] = [];
		for (const layer of SoundLayer.layers) {
			names.push(layer.value!);
		}
		return names;
	}

	/**
	 * Retrieves sound layer corresponding to layer name.
	 *
	 * @param name {string}
	 *   Layer name.
	 * @return {data.sound.SoundLayer.SoundLayer}
	 *   Sound layer matching name or `undefined`.
	 */
	static checkLayerName(name: string): SoundLayer|undefined {
		for (const layer of SoundLayer.layers) {
			if (layer.value === name) {
				return layer;
			}
		}
		return undefined;
	}
}
